src/libmain/progress-bar.cc

```diff
@@ -86,6 +86,9 @@ class ProgressBar : public Logger
     bool printBuildLogs = false;
     bool isTTY;
 
+    /** Helps avoid unnecessary redraws, see `draw()` */
+    std::string lastOutput;
```

Comment:
Shouldn't this be in `State`? Maybe it's implicitly protected by the `State` mutex right now, but it's best not to rely on that.